src/lib/openapi/index.ts

```diff
@@ -147,8 +147,36 @@ import { clientMetricsEnvSchema } from './spec/client-metrics-env-schema';
 import { updateTagsSchema } from './spec/update-tags-schema';
 import { batchStaleSchema } from './spec/batch-stale-schema';
 
+// Schemas must have an $id property on the form "#/components/schemas/mySchema".
+export type SchemaId = typeof schemas[keyof typeof schemas]['$id'];
+
+// Schemas must list all their $refs in `components`, including nested schemas.
+export type SchemaRef = typeof schemas[keyof typeof schemas]['components'];
+
+// JSON schema properties that should not be included in the OpenAPI spec.
+export interface JsonSchemaProps {
+    $id: string;
+    components: object;
+}
+
+type SchemaWithMandatoryFields = Partial<
+    Omit<
+        OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject,
+        '$id' | 'components'
+    >
+> &
+    JsonSchemaProps;
+
+interface UnleashSchemas {
+    [name: string]: SchemaWithMandatoryFields;
+}
+
+interface OpenAPIV3DocumentWithServers extends OpenAPIV3.Document {
+    servers: OpenAPIV3.ServerObject[];
+}
+
 // All schemas in `openapi/spec` should be listed here.
-export const schemas = {
+export const schemas: UnleashSchemas = {
```

Comment:
This just forces some validations at compilation time, giving us a faster feedback loop